### PR TITLE
MapExplorer - May 2016 improvements [Rebased][READY FOR REVIEW]

### PR DESCRIPTION
**Pull request by shaunanoordin**
Let's take another stab at the MapExplorer
- We want a UI that's cleaner and easier to use than the existing version
- We want to leverage on some of CartoDB's Deep Insights functionality
- We want to remove all the unused stuff that's been hanging around since the initial alpha. (Plus code refactor.)
- Maybe... we want to reorganise the component names and file-folder structure, for orderliness' sake?

### CODE REVIEW

**Comment by shaunanoordin** [comment]
screen shot 2016-05-19 at 15 37 31
screen shot 2016-05-19 at 15 37 39

Current state of the experiment/improvements:
- Check it out - the Selectors/Filters are now a slide-in, slide-out menu to provide the Selectors and Map Visuals primo screen estate when they're at the fore.
- The MapVisualisations can now be logically separated from the Map Explorer, making it easier to plug & play different mapping systems in the future. That's my next ToDo

I won't lie, I'm kinda afraid that this might turn into another mega-PR like what old me used to do, but I can't imagine merging this PR until the experiment _(which would be the mega-change)_ is stabilised. If you have advice, I'm all ears.

**Comment by shaunanoordin** [comment]
FYI the Map Explorer and the Map Visuals have been split; this means that the actual visualisation of the Map Data is a separate sub-component in the MapExplorer system.

File: src/containers/MapExplorer-SelectorPanel.jsx [hunk 1]
```diff
@@ -72,16 +70,14 @@ class SelectorPanel extends Component {
...
+          <div className="input-row">
             <label>SPECIES:</label>
             {
               (speciesText.length > 0)
               ? <span>Viewing {speciesText}</span>
               : <span>Viewing all species</span>
             }
-            <button onClick={(e) => { this.setState({ speciesDialog: { status: DialogScreen.DIALOG_ACTIVE } }) }}>...</button>
           </div>
-          <DialogScreen_Species ref="dialog_species" status={this.state.speciesDialog.status} data={this.props.selectorData.species} updateMeHandler={this.updateSpeciesSelection} closeMeHandler={this.closeAllDialogs} />
```
Comment (shaunanoordin) [inline src/containers/MapExplorer-SelectorPanel.jsx:84]: This old experiment has been phased out.

File: src/containers/MapExplorer-SelectorPanel.jsx [hunk 2]
```diff
@@ -269,12 +265,6 @@ class SelectorPanel extends Component {
   
   //----------------------------------------------------------------
   
-  updateSpeciesSelection(e) {
```
Comment (shaunanoordin) [inline src/containers/MapExplorer-SelectorPanel.jsx]: Old experiment, phased out

File: src/containers/MapExplorer-SelectorPanel.jsx [hunk 3]
```diff
@@ -133,7 +129,7 @@ class SelectorPanel extends Component {
           </div>
         </section>
         <section className="action-subpanel">
-          <button className="btn" onClick={this.updateMe}>(Apply)</button>
+          <button className="hidden" onClick={this.updateMe}>(Apply)</button>
```
Comment (shaunanoordin) [inline src/containers/MapExplorer-SelectorPanel.jsx:132]: Slowly replacing "edit this one filter selection" option with a "edit ALL filter selections" feature

File: src/containers/MapExplorer-SelectorPanel.jsx [hunk 4]
```diff
@@ -340,4 +330,7 @@ class SelectorPanel extends Component {
 SelectorPanel.propTypes = {
   dispatch: PropTypes.func.isRequired
 };
-export default connect()(SelectorPanel);
+
+//Don't subscribe to the Redux Store, but gain access to dispatch() and give
+//this component's parent access to this component via getWrappedInstance()
+export default connect(null, null, null, { withRef: true })(SelectorPanel);
```
Comment (shaunanoordin) [inline src/containers/MapExplorer-SelectorPanel.jsx:336]: Useful note on this: the `{withRef:true}` option lets an external component (e.g. the parent) access the current component, even after the component has been wrapped up by a `connect`.

**Comment by shaunanoordin** [comment]
The Map Explorer is going through quite a massive UI/UX change, but I wanted to stop here and make a 'checkpoint' to ensure my ideas are on the right track. And so my PR doesn't become even MOAR MEGA HUEG. @simoneduca @eatyourgreens @aliburchard, I'm keen to hear your comments.

TL;DR:
- Map Explorer now has a sliding panel that shows more controls.
- The code has been cleaned up and made less nonsensical.

UI/UX:
- The Map Explorer now sports a larger screen estate for controls; this will allow us to insert control widgets to, e.g. better control & visualise distance to humans/water, have a more verbose & friendly species selection, etc.
- I'm thinking of _moving away_ from the "multiple selectors" architecture and focusing on having only one selector... i.e. stick with the way it's been working so far, since nobody has expressed a need to have multiple selectors _yet._ (e.g. nobody said "I want yellow spots to highlight lions, red spots to highlight herbivores)
- Also, there's a new "centre map on WildCam Gorongosa" button on the map, on the top left, marked with a crosshair

Code:
- The biggest refactor is that the Map visualisation code inside MapExplorer has been split off into a MapVisuals component. I like this.
- Next time I'll take a look at compartmentalising the map controls into, oh, MapControls and cleaning up the 'logic' side of the MapExplorer. (e.g. shunting MapExplorer-SelectorData into a new folder called... `/data-structures`... because it's not really a `container` now, is it)

screen shot 2016-05-26 at 18 13 11

screen shot 2016-05-26 at 18 13 18

**Comment by shaunanoordin** [comment]
...I'll wait until #219 gets merged before I rebase this PR.

**Comment by shaunanoordin** [comment]
Bam! Rebased!

**Comment by shaunanoordin** [comment]
Tested on IE10 and Edge in addition to the usual non-insane browsers.

File: src/styles/components/dialog-screen.styl [hunk 5]
```diff
@@ -2,7 +2,7 @@
   position: fixed
   left: 0
   top: 0
-  z-index: 1000000  //Place waaay above map plugin elements.
+  z-index: 3000000  //Place waaay above map plugin elements.
```
Comment (simoneduca) [inline src/styles/components/dialog-screen.styl]: @shaunanoordin (not a blocker) Why is such a big number needed? What issue is increasing number solving? We don't use `z-index` anywhere else, do we? If a high number is needed, we could follow some kind of convention like https://css-tricks.com/rational-z-index-values/ or https://css-tricks.com/handling-z-index/

Comment (shaunanoordin) [inline src/styles/components/dialog-screen.styl]: I absolutely agree with the idea of having 'rational' z-index values. I'd like to keep the z-index low with each layer at a multiple of 10 or 100, but as I discovered...

screen shot 2016-05-31 at 17 03 36

...the CartoDB logo itself comes in at a staggering **1-million** z-index.

Dr Evil's One Million Dollars meme

Admittedly, my 3-million z-index was an overreaction - `z-index: 1,000,001 would be more than sufficient at handling the layering.

**Comment by shaunanoordin** [comment]
@simoneduca I've improved the z-index value of the popup dialogs and added a close button! :)

**Comment by simoneduca** [comment]
Nice one, LGTM!
